esp/esp/program/modules/handlers/studentregphasezeromanage.py

```diff
@@ -0,0 +1,161 @@
+
+        #Get lottery records and randomize them
+        records = PhaseZeroRecord.objects.filter(program=prog).order_by('?')
+
+        ###############################################################################
+        # The lottery algorithm is run, with randomization and processing in order.
+        # If any one in the group doesn't get in (due to cap size), no one in that group gets in.
+        counts = {key:0 for key in grade_caps}
+        winners = Group.objects.get_or_create(name=role)[0]
+
+        for i in records:
+            sibs = i.user.all()
+            newcounts = copy.copy(counts)
+            for j in sibs:
+                newcounts[j.getGrade(prog)] += 1
+
+            cpass = True
+            for c in counts.keys():
+                if newcounts[c] > grade_caps[c]:
+                    cpass = False
+
+            if cpass:
+                for user in sibs:
+                    user.groups.add(winners)
+                counts = copy.copy(newcounts)
+
+        ###############################################################################
+        # Post lottery, assign permissions to people in the lottery winners group
+        # Assign OverridePhaseZero permission and Student/All permissions
+
+        override_perm = Permission(permission_type='OverridePhaseZero', role=winners, start_date=datetime.datetime.now(), program=prog)
+        studentAll_perm = Permission(permission_type='Student/All', role=winners, start_date=datetime.datetime.now(), program=prog)
+        override_perm.save()
+        studentAll_perm.save()
+        # Add tag to indicate student lottery has been run
+        Tag.setTag('student_lottery_run', target=prog, value='True')
+
+        return True
+
+    @main_call
+    @needs_admin
+    def phasezero(self, request, tl, one, two, module, extra, prog):
+        context = {}
+        role = str(prog) + " Winners"
+        context['role'] = role
+        q_phasezero = Q(phasezerorecord__program=self.program)
+        entrants = ESPUser.objects.filter(q_phasezero).distinct()
+        context['entrants'] = entrants
+        context['nentrants'] = len(entrants)
+
+        grades = range(prog.grade_min, prog.grade_max + 1)
+        stats = {}
+
+        #Calculate grade counts
+        for grade in grades:
+            stats[grade] = {}
+            stats[grade]['in_lottery'] = 0
+        for entrant in entrants:
+            stats[entrant.getGrade(prog)]['in_lottery'] += 1
+
+        #Run lottery if requested
+        if request.POST:
+            if Tag.getBooleanTag('student_lottery_run', prog, default=False):
+                context['error'] = "You've already run the student lottery!"
+            else:
+                role = request.POST['rolename']
+                context['role'] = role
+                if "confirm" in request.POST:
+                    success = False
+                    success = self.lottery(prog, role)
```

Comment:
Can this ever not return `True`?  If you want to catch errors, better to do `try`/`except`.